src/pls.jl

```diff
@@ -217,10 +225,35 @@ function fitted!(v::AbstractArray{T}, m::LinearMixedModel{T}) where T
     v
 end
 
+fitted!(v::AbstractArray{T}, m::LinearMixedModel{T}) where {T} = fitted!(v, m, m.trms)
+
 StatsBase.fitted(m::LinearMixedModel{T}) where {T} = fitted!(Vector{T}(nobs(m)), m)
 
 StatsBase.predict(m::LinearMixedModel) = fitted(m)
 
+function StatsBase.predict(m::LinearMixedModel{T}, test::AbstractDataFrame; #, train::AbstractDataFrame;
+	 contrasts=Dict()) where T
+    mf = ModelFrame(m.formula, test, contrasts=contrasts)
+    X = ModelMatrix(mf).m
+    n = size(X, 1)
+    T = eltype(X)
+
+    # # can the tdict be used from the original data? (without building this modelframe again)
+    # mf_train = ModelFrame(m.formula, train, contrasts=contrasts)    
+    # tdict = get_tdict(mf_train.terms.terms)
+    tdict = get_tdict(mf.terms.terms)
+
+    # make a new trms object
+    trms = get_trms(tdict,mf.df,test,T,n)
+    
+    sort!(trms, by = nrandomeff, rev = true)
```

Comment:
I don't think you want to sort the `trms` according to `nrandomeff` in the new data frame.

You probably want to align the random effects terms according to the grouping factor which is stored as the `fnm` member of the struct.

Reply:
Aha! This could be the source of one of the issues I am experiences. I'll take a look at this.

Reply:
Okay I've answered my part 1 question: yes it uses the data.

Reply:
I'm not sure why they need to be sorted in the first place (I suspect there is a reason, although it's also fine if you don't want to explain it to me).

Reply:
Maybe they were sorted explicitly because the old code used the `refs` field of `CategoricalArray`, which isn't sorted? If so, using `order` would make it unnecessary to sort explicitly (which breaks any user-defined order).

Reply:
Hm, I don't think so. This is sorting each different random effect, not for within an effect.

I fixed the code that I wrote to respect the original sort, anyway.